pennylane_lightning/core/src/simulators/lightning_qubit/gates/GateIndices.cpp

```diff
@@ -47,4 +47,34 @@ auto generateBitPatterns(const std::vector<std::size_t> &qubitIndices,
     }
     return indices;
 }
+
+/**
+ * @brief Introduce quantum controls in indices generated by
+ * generateBitPatterns.
+ *
+ * @param indices Indices for the operation.
+ * @param num_qubits Number of qubits in register.
+ * @param controlled_wires Control wires.
+ * @param controlled_values Control values (false or true).
+ */
+void controlBitPatterns(std::vector<std::size_t> &indices,
+                        const std::size_t num_qubits,
+                        const std::vector<std::size_t> &controlled_wires,
+                        const std::vector<bool> &controlled_values) {
+    constexpr std::size_t one{1U};
+    if (controlled_wires.empty()) {
+        return;
+    }
+    std::for_each(
+        indices.begin(), indices.end(),
+        [num_qubits, &controlled_wires, &controlled_values](std::size_t &i) {
```

Comment:
Minor:
```suggestion
        [&num_qubits, &controlled_wires, &controlled_values](std::size_t &i) {
```

Reply:
No harm in that, but in our C++ guidelines we say copy POD (plain old data) objects.

Reply:
It doesn't include `std::size_t` values, does it? :smile:

Reply:
Not be the right guideline then. Maybe it does not apply to lambdas, but in a regular functions I would not pass such a small object by reference because of the overhead reading from the stack. For example, throw 
```
#include <cstddef>

std::size_t foo(std::size_t& x)
{
    return x*x;
}

std::size_t bar(std::size_t x)
{
    return x*x;
}
```
into [godbolt](https://godbolt.org/) with GCC and -O2/3.

Reply:
Yeah, nothing with regular functions! Although copying a `std::size_t` variable is very cheap, in this function as the lambda will be invoked many times in the critical section and  the lambda is copied multiple times, the perf difference between passing by reference and by value may be still negligible but in favour of passing by reference.